File: core/src/test/java/me/raatiniemi/worker/domain/interactor/GetProjectTimeSinceTest.java
/*
 * Copyright (C) 2017 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.domain.interactor;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Calendar;

import me.raatiniemi.worker.domain.exception.InvalidStartingPointException;
import me.raatiniemi.worker.domain.model.Project;
import me.raatiniemi.worker.domain.repository.TimeIntervalRepository;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith(JUnit4.class)
public class GetProjectTimeSinceTest {
    private TimeIntervalRepository timeIntervalRepository;
    private GetProjectTimeSince getProjectTimeSince;
    private Project project;

    private static long getMillisecondsForStartingPoint(int startingPoint) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        switch (startingPoint) {
            case GetProjectTimeSince.DAY:
                break;
            case GetProjectTimeSince.WEEK:
                calendar.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
                break;
            case GetProjectTimeSince.MONTH:
                calendar.set(Calendar.DAY_OF_MONTH, 1);
                break;
            default:
                throw new InvalidStartingPointException(
                        "Starting point '" + startingPoint + "' is not valid"
                );
        }

        return calendar.getTimeInMillis();
    }

    @Before
    public void setUp() {
        timeIntervalRepository = mock(TimeIntervalRepository.class);
        getProjectTimeSince = new GetProjectTimeSince(timeIntervalRepository);
        project = Project.builder("Name")
                .id(1L)
                .build();
    }

    @Test
    public void execute_withDay() {
        getProjectTimeSince.execute(project, GetProjectTimeSince.DAY);

        verify(timeIntervalRepository)
                .findAll(
                        eq(project),
                        eq(getMillisecondsForStartingPoint(GetProjectTimeSince.DAY))
                );
    }

    @Test
    public void execute_withWeek() {
        getProjectTimeSince.execute(project, GetProjectTimeSince.WEEK);

        verify(timeIntervalRepository)
                .findAll(
                        eq(project),
                        eq(getMillisecondsForStartingPoint(GetProjectTimeSince.WEEK))
                );
    }

    @Test
    public void execute_withMonth() {
        getProjectTimeSince.execute(project, GetProjectTimeSince.MONTH);

        verify(timeIntervalRepository)
                .findAll(
                        eq(project),
                        eq(getMillisecondsForStartingPoint(GetProjectTimeSince.MONTH))
                );
    }

    @Test(expected = InvalidStartingPointException.class)
    public void execute_withInvalidStartingPoint() {
        getProjectTimeSince.execute(project, -1);
    }
}

File: core/src/test/java/me/raatiniemi/worker/domain/interactor/GetProjectTimeSinceTest.kt
/*
 * Copyright (C) 2018 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.domain.interactor

import me.raatiniemi.worker.domain.date.hours
import me.raatiniemi.worker.domain.date.minutes
import me.raatiniemi.worker.domain.model.*
import me.raatiniemi.worker.domain.repository.TimeIntervalInMemoryRepository
import me.raatiniemi.worker.domain.repository.TimeIntervalRepository
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class GetProjectTimeSinceTest {
    private val project = Project(1, "Name")

    private lateinit var repository: TimeIntervalRepository
    private lateinit var getProjectTimeSince: GetProjectTimeSince

    @Before
    fun setUp() {
        repository = TimeIntervalInMemoryRepository()
        getProjectTimeSince = GetProjectTimeSince(repository)
    }

    @Test
    fun `get project time since day`() {
        val newTimeInterval = newTimeInterval {
            start = after(TimeIntervalStartingPoint.DAY)
            stop = after(TimeIntervalStartingPoint.DAY, 2.hours)
        }
        repository.add(
            newTimeInterval {
                start = before(TimeIntervalStartingPoint.DAY)
                stop = before(TimeIntervalStartingPoint.DAY, 30.minutes)
            }
        )
        repository.add(newTimeInterval)
        val expected = listOf(
            timeInterval {
                id = 2
                start = newTimeInterval.start
                stop = requireNotNull(newTimeInterval.stop)
            }
        )

        val actual = getProjectTimeSince(project, TimeIntervalStartingPoint.DAY)

        assertEquals(expected, actual)
    }

    @Test
    fun `get project time since week`() {
        val newTimeInterval = newTimeInterval {
            start = after(TimeIntervalStartingPoint.WEEK)
            stop = after(TimeIntervalStartingPoint.WEEK, 2.hours)
        }
        repository.add(
            newTimeInterval {
                start = before(TimeIntervalStartingPoint.WEEK)
                stop = before(TimeIntervalStartingPoint.WEEK, 30.minutes)
            }
        )
        repository.add(newTimeInterval)
        val expected = listOf(
            timeInterval {
                id = 2
                start = newTimeInterval.start
                stop = requireNotNull(newTimeInterval.stop)
            }
        )

        val actual = getProjectTimeSince(project, TimeIntervalStartingPoint.WEEK)

        assertEquals(expected, actual)
    }

    @Test
    fun `get project time since month`() {
        val newTimeInterval = newTimeInterval {
            start = after(TimeIntervalStartingPoint.MONTH)
            stop = after(TimeIntervalStartingPoint.MONTH, 2.hours)
        }
        repository.add(
            newTimeInterval {
                start = before(TimeIntervalStartingPoint.MONTH)
                stop = before(TimeIntervalStartingPoint.MONTH, 30.minutes)
            }
        )
        repository.add(newTimeInterval)
        val expected = listOf(
            timeInterval {
                id = 2
                start = newTimeInterval.start
                stop = requireNotNull(newTimeInterval.stop)
            }
        )

        val actual = getProjectTimeSince(project, TimeIntervalStartingPoint.MONTH)

        assertEquals(expected, actual)
    }
}
